File: Algorithms/Python/63 Unique Paths II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        dp = [1] * n
        for i in range(n):
            # encounter obstacle
            if obstacleGrid[0][i] == 1 or (i > 0 and dp[i - 1] == 0):
                dp[i] = 0
        
        for i in range(1, m):
            for j in range(n):
                if obstacleGrid[i][j] == 1 or (j == 0 and dp[j] == 0):
                    dp[j] = 0
                else:
                    if j == 0:
                        dp[j] = 1
                    else:
                        dp[j] = dp[j - 1] + dp[j]

        return dp[n - 1]

File: Algorithms/C/63 Unique Paths II.c
int uniquePathsWithObstacles(int** obstacleGrid, int obstacleGridSize, int* obstacleGridColSize){
    int* dp = (int*)malloc(sizeof(int)*(*obstacleGridColSize));
    for (int i = 0; i < *obstacleGridColSize; ++i){
        if (obstacleGrid[0][i] == 1 || (i > 0 && dp[i - 1] == 0)){
            dp[i] = 0;
        }
        else{
            dp[i] = 1;
        }
    }

    for (int i = 1; i < obstacleGridSize; ++i)
        for (int j = 0; j < *obstacleGridColSize; ++j){
            if (obstacleGrid[i][j] == 1 || (j == 0 && dp[j] == 0)){
                dp[j] = 0;
            }
            else{
                if (j == 0){
                    dp[j] = 1;
                }
                else{
                    dp[j] = dp[j] + dp[j - 1];
                }
            }
        }

    int result = dp[*obstacleGridColSize - 1];
    free(dp);
    return result;
}
